kiwmi/desktop/output.c

```diff
@@ -376,8 +388,143 @@ new_output_notify(struct wl_listener *listener, void *data)
     wl_signal_emit(&desktop->events.new_output, output);
 }
 
+void
+output_layout_change_notify(struct wl_listener *listener, void *UNUSED(data))
+{
+    struct kiwmi_desktop *desktop =
+        wl_container_of(listener, desktop, output_layout_change);
+    struct wlr_output_configuration_v1 *configuration =
+        wlr_output_configuration_v1_create();
+
+    struct kiwmi_output *output;
+    wl_list_for_each (output, &desktop->outputs, link) {
+        struct wlr_output_configuration_head_v1 *head =
+            wlr_output_configuration_head_v1_create(
+                configuration, output->wlr_output);
+
+        struct wlr_box *area = wlr_output_layout_get_box(
+            desktop->output_layout, output->wlr_output);
+
+        if (area != NULL) {
+            head->state.x = area->x;
+            head->state.y = area->y;
+        }
+    }
+
+    wlr_output_manager_v1_set_configuration(
+        desktop->output_manager, configuration);
+}
+
+static void
+output_manager_configure(
+    struct kiwmi_server *server,
+    struct wlr_output_configuration_v1 *configuration,
+    bool test)
+{
+    struct wlr_output_configuration_head_v1 *head;
+    struct wlr_output_configuration_head_v1 *bad_head = NULL;
+    wl_list_for_each (head, &configuration->heads, link) {
+        struct wlr_output *wlr_output = head->state.output;
+
+        wlr_output_enable(wlr_output, head->state.enabled);
+        if (head->state.enabled) {
+            if (head->state.mode) {
+                wlr_output_set_mode(wlr_output, head->state.mode);
+            } else {
+                wlr_output_set_custom_mode(
+                    wlr_output,
+                    head->state.custom_mode.width,
+                    head->state.custom_mode.height,
+                    head->state.custom_mode.refresh);
+            }
+
+            wlr_output_set_transform(wlr_output, head->state.transform);
+            wlr_output_set_scale(wlr_output, head->state.scale);
+        }
+
+        if (!wlr_output_test(wlr_output)) {
+            bad_head = head;
+            break;
+        }
+    }
+
+    wl_list_for_each (head, &configuration->heads, link) {
+        if (bad_head == NULL && !test) {
```

Comment:
This seems like a ‘special-case’ (or whatevs) condition with a short & a long branch. It could help readability to invert it and `continue` at the end of what’s now the `else` branch, but i’ll leave that decision to you.

Reply:
Fair enough, saves an indentation level at least.